File: zoo/src/main/scala/com/intel/analytics/zoo/models/transformer/Transformer.scala
/*
 * Copyright 2018 Analytics Zoo Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.intel.analytics.zoo.models.transformer

import com.intel.analytics.bigdl.nn.abstractnn.{AbstractModule, Activity}
import com.intel.analytics.bigdl.nn.keras.KerasLayer
import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
import com.intel.analytics.bigdl.utils.Shape
import com.intel.analytics.zoo.models.common.ZooModel
import com.intel.analytics.zoo.pipeline.api.keras.layers._
import com.intel.analytics.zoo.pipeline.api.autograd.{AutoGrad, Parameter, Variable}
import com.intel.analytics.zoo.pipeline.api.keras.models.Model
import com.intel.analytics.bigdl.tensor.Tensor

import scala.reflect.ClassTag

class Transformer[T: ClassTag] private(
  vocab: Int,
  nCtx: Int,
  embeddingSize: Int,
  embeddingDrop: Double,
  nLayer: Int,
//  afn: String,
  residPdrop: Double,
  attnPdrop: Double,
  nHead: Int)(implicit ev: TensorNumeric[T])
  extends ZooModel[Activity, Activity, T] {

  override def buildModel(): AbstractModule[Activity, Activity, T] = {
    // (sequence length, position)
    val input = Variable(Shape(nCtx, 2))
    val r = Reshape(Array(nCtx * 2)).from(input)
    val e = Embedding(vocab, embeddingSize, inputLength = nCtx * 2).from(r)

    val r2 = Reshape(Array(nCtx, 2, embeddingSize)).from(e)

//  Add the position information to the input embeddings
    val h = AutoGrad.sum(r2, 2, false)

    var nextInput: Variable[T] = h

    for (i <- 0 until nLayer) {
      val output = block(nextInput)
      nextInput = output
    }

    val model = Model(input, nextInput)
    model
  }

  def block(x: Variable[T]): Variable[T] = {
    val a = attention(x)
    val n = layerNorm(x + a)
    val m = mlp(n)
    val h = layerNorm2(n + m)
    h
  }

  // g, b for layerNorm
  val g = Parameter[T](Shape(embeddingSize), initWeight = Tensor.ones(embeddingSize))
  val b = Parameter[T](Shape(embeddingSize), initWeight = Tensor(embeddingSize))
  def layerNorm(x: Variable[T], e: Double = 1e-5): Variable[T] = {
    val sizes = x.getOutputShape().toSingle().toArray
    val u = AutoGrad.mean(x, sizes.size - 1, true)
    val s = AutoGrad.mean(AutoGrad.square(x - u), sizes.size - 1, true)
    val y = (x - u) / AutoGrad.sqrt(s + e)
    TimeDistributed[T](Dense[T](embeddingSize)
      .asInstanceOf[KerasLayer[Activity, Tensor[T], T]]).from(y)
  }

  def layerNorm2(x: Variable[T], e: Double = 1e-5): Variable[T] = {
    val sizes = x.getOutputShape().toSingle().toArray
    val u = AutoGrad.mean(x, sizes.size - 1, true)
    val s = AutoGrad.mean(AutoGrad.square(x - u), sizes.size -1, true)
    val y = (x - u) / AutoGrad.sqrt(s + e)
    TimeDistributed[T](Dense[T](embeddingSize)
      .asInstanceOf[KerasLayer[Activity, Tensor[T], T]]).from(y)
  }

  def gelu(x: Variable[T]): Variable[T] = {
    x * 0.5 * (Activation("tanh").from((AutoGrad.square(x) * x * 0.044715 + x)
      * (scala.math.sqrt(2 / scala.math.Pi))) + 1)
  }

  def mlp(x: Variable[T]): Variable[T] = {
    val h = Conv1D(embeddingSize * 4, 1).from(x)
    val a = gelu(h)
    val h2 = Conv1D(embeddingSize, 1).from(a)
    Dropout(residPdrop).from(h2)
  }

  def attention(x: Variable[T]): Variable[T] = {
    val c = Conv1D(embeddingSize * 3, 1).from(x)
    val query = c.slice(2, 0, embeddingSize)
    val key = c.slice(2, embeddingSize, embeddingSize)
    val value = c.slice(2, embeddingSize * 2, embeddingSize)
    val q = splitHeads(query)
    val k = splitHeads(key, k = true)
    val v = splitHeads(value)
    val a = attn(q, k, v) // m: (-1, 12, 77, 64)
    val m = mergeHeads(a) // m: (-1, 77, 768)
    val n = Conv1D(embeddingSize, 1).from(m) // n: (-1, 77, 768)
    Dropout(residPdrop).from(n)
  }

  def splitHeads(x: Variable[T], k: Boolean = false): Variable[T] = {
    val sizes = x.getOutputShape().toSingle().toArray
    val newSizes = sizes.drop(1).dropRight(1) ++ Array(nHead, sizes.last / nHead)
    val r = Reshape(newSizes).from(x)
    if (k) Permute(Array(2, 3, 1)).from(r)
    else Permute(Array(2, 1, 3)).from(r)
  }

  def mergeHeads(x: Variable[T]): Variable[T] = {
    val p = AutoGrad.contiguous(Permute[T](Array(2, 1, 3)).from(x))
    val sizes = p.getOutputShape().toSingle().toArray
    Reshape(sizes.drop(1).dropRight(2) ++ Array(sizes.last * sizes(sizes.length - 2))).from(p)
  }

  // weights and ab belong to Attention
  val weights = Utils.tril(Tensor.ones(nCtx, nCtx)).view(1, 77, 77)
  val ab = Parameter[T](Shape(1, nCtx, nCtx), trainable = false, initWeight = weights)

  // scale shoule be set in Attention
  def attn(q: Variable[T], k: Variable[T], v: Variable[T], scale: Boolean = false): Variable[T] = {
    // q:(16, 12, 77, 64) k:(16, 12, 64, 77) v:(16, 12, 77, 64)
    var w = AutoGrad.mm(q, k) // w: (16, 12, 77, 77)
    if (scale) w = w / scala.math.sqrt(v.getOutputShape().toSingle().toArray.last)

    w = AutoGrad.mm(ab, w) + (ab * (-1) + 1) * -1e9
    // TODO: softmax with last dim
//    val lw = w.indexSelect(-1, 0)
    w = Activation("softmax").from(w)
    w = Dropout(attnPdrop).from(w)

    AutoGrad.mm(w, v)
  }
}

object Transformer {
  def apply[@specialized(Float, Double) T: ClassTag](
    vocab: Int = 40990,
    nCtx: Int = 77, // seq len
    embeddingSize: Int = 768,
    embeddingDrop: Double = 0.1,
    nLayer: Int = 12,
//    afn: String = "gelu",
    residPdrop: Double = 0.1,
    attnPdrop: Double = 0.1,
    nHead: Int = 12)(implicit ev: TensorNumeric[T]): Transformer[T] = {
    new Transformer[T](vocab, nCtx, embeddingSize, embeddingDrop, nLayer,
//      afn,
      residPdrop, attnPdrop, nHead).build()
  }

  /**
    * Load an existing Transformer model (with weights).
    *
    * @param path The path for the pre-defined model.
    *             Local file system, HDFS and Amazon S3 are supported.
    *             HDFS path should be like "hdfs://[host]:[port]/xxx".
    *             Amazon S3 path should be like "s3a://bucket/xxx".
    * @param weightPath The path for pre-trained weights if any. Default is null.
    * @tparam T Numeric type of parameter(e.g. weight, bias). Only support float/double now.
    */
  def loadModel[T: ClassTag](
    path: String,
    weightPath: String = null)(implicit ev: TensorNumeric[T]): Transformer[T] = {
    ZooModel.loadModel(path, weightPath).asInstanceOf[Transformer[T]]
  }
}

File: pyzoo/zoo/pipeline/api/keras/layers/transformer.py
#
# Copyright 2018 Analytics Zoo Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

from bigdl.nn.layer import Sum

from zoo.pipeline.api.keras.engine import ZooKerasLayer
from zoo.pipeline.api.keras.layers import *
from zoo.pipeline.api.keras.models import Sequential


if sys.version >= '3':
    long = int
    unicode = str


class TransformerLayer(ZooKerasLayer):
    """
    A self attention layer

    # Arguments
    nBlock: block number
    resid_drop: drop probability off projection
    attn_drop: drop probability of attention
    n_head: head number
    mask_attention: whether unidirectional or bidirectional
    embedding_layer: embedding layer
    >>> layer = TransformerLayer.init_with_DefaultEmbedding()
    creating: createZooKerasSequential
    creating: createZooKerasReshape
    creating: createZooKerasEmbedding
    creating: createZooKerasDropout
    creating: createZooKerasReshape
    creating: createSum
    creating: createZooKerasKerasLayerWrapper
    creating: createSqueeze
    creating: createZooKerasKerasLayerWrapper
    creating: createZooKerasTransformerLayer

    """
    def __init__(self, n_block, resid_drop, attn_drop,
                 n_head, mask_attention, embedding_layer):
        super(TransformerLayer, self).__init__(None, n_block, resid_drop, attn_drop, n_head,
                                               mask_attention, embedding_layer)

    @classmethod
    def init_with_default_embedding(cls, vocab=40990, seq_len=77, n_block=12, resid_drop=0.1,
                                    attn_drop=0.1, n_head=12, embedding_size=768,
                                    embedding_drop=0.1, mask_attention=True):
        """
        vocab: vocabulary size of training data, default is 40990
        seq_len: max sequence length of training data, default is 77
        n_block: block number, default is 12
        resid_drop: drop probability of projection, default is 0.1
        attn_drop: drop probability of attention, default is 0.1
        n_head: head number, default is 12
        embedding_size: embedding size
        embedding_drop: drop probability of embedding layer, default is 0.1
        mask_attention: whether unidirectional or bidirectional, default is true(unidirectional)
        """
        from bigdl.nn.layer import Squeeze
        embedding = Sequential()

        embedding.add(Reshape([seq_len * 2], input_shape=(seq_len, 2)))\
            .add(Embedding(vocab, embedding_size, input_length=seq_len * 2))\
            .add(Dropout(embedding_drop))\
            .add(Reshape((seq_len, 2, embedding_size)))\
            .add(KerasLayerWrapper(Sum(dimension=3, squeeze=True)))
        # walk around for bug #1208, need remove this line after the bug fixed
        embedding.add(KerasLayerWrapper(Squeeze(dim=3)))

        return TransformerLayer(n_block, resid_drop, attn_drop, n_head, mask_attention, embedding)
